### PR TITLE
PFW-1536: Reprint function implemented for MK3S/MK3S+

### PR DESCRIPTION
**Pull request by jfestrada**
This PR is a the same than #3071 but using the current repository of my fork of your firmware

To support this functionality I have added a **extern bool variable** which can be read/set from cmdqueue.cpp and **ultralcd.cpp** so when the card reader reaches the EOF the variable isPrintFinished will be set to true and the main menu will show the **REPRINT** option.
How do I know which file we need to reprint? All the information needed to perform the reprint was stored on the EEPROM in case of a power panic and the fw recover this information in order to start the print from SD again, so using that information the function reprint_from_eeprom() has been implemented.

Translations:
- [X] :czech_republic: [same as buddy firmware](https://github.com/prusa3d/Prusa-Firmware-Buddy/blob/bfb0ffc745ee6546e7efdba618d0e7c0f4c909cd/src/lang/po/cs/Prusa-Firmware-Buddy_cs.po#L3647-L3648)
- [X] :de: [same as buddy firmware](https://github.com/prusa3d/Prusa-Firmware-Buddy/blob/bfb0ffc745ee6546e7efdba618d0e7c0f4c909cd/src/lang/po/de/Prusa-Firmware-Buddy_de.po#L3737-L3738)
- [X] :es: [same as buddy firmware](https://github.com/prusa3d/Prusa-Firmware-Buddy/blob/bfb0ffc745ee6546e7efdba618d0e7c0f4c909cd/src/lang/po/es/Prusa-Firmware-Buddy_es.po#L3710-L3711)
- [X] :fr: [same as buddy firmware](https://github.com/prusa3d/Prusa-Firmware-Buddy/blob/bfb0ffc745ee6546e7efdba618d0e7c0f4c909cd/src/lang/po/fr/Prusa-Firmware-Buddy_fr.po#L3676-L3677)
- [X] :it: [same as buddy firmware](https://github.com/prusa3d/Prusa-Firmware-Buddy/blob/bfb0ffc745ee6546e7efdba618d0e7c0f4c909cd/src/lang/po/it/Prusa-Firmware-Buddy_it.po#L3698-L3699)
- [X] :poland: [same as buddy firmware](https://github.com/prusa3d/Prusa-Firmware-Buddy/blob/bfb0ffc745ee6546e7efdba618d0e7c0f4c909cd/src/lang/po/pl/Prusa-Firmware-Buddy_pl.po#L3679-L3680)
- [X] :netherlands: by @3d-gussner @vintagepc please check [Firmware_nl.po](https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-92de134d7cfe75b62e86a4e83716cd95550b11f2ddee228150a9c6faa4d6a34bR2585-R2586)
- [ ] :croatia: @Prime1910 please check [Firmware_hr.po](https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-487d5469566575627c7dc47abf437c52e890de941c1fd8d4db5a2a4d6ef08d64R2577-R2578)
- [x] :hungary: @AttilaSVK please check [Firmware_hu.po](https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-3a7d51d07fb16c60d273b7986aa88a7892939b19a9bad4f3501a153ffb0604e9R2582-R2583)
- [ ] :norway: @OS-kar and @trondkla please check [Firmware_no.po](https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-22e5dd5d54111e837f3b40db65f2a126b09cc4157dcfa1f34b6e5dde0bf4a607R2559-R2560)
- [x] :romania: @leptun and @Hauzman please check [Firmware_ro.po](https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-6774f424618c20ccb1d99bf93cb34bab0b6ee03a55868fa7021a11ef6c28667fR2583-R2584)
- [x] :slovakia: @ingbrzy and @shatter136 please check [Firmware_sk.po](https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-9584ba5c6a2b7618c16dc96fa21d692e9d5d69c393c55f37166bb05e98557fbcR2565-R2566)
- [ ] :sweden: @Painkiller56 please check [Firmware_sv.po](https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-cf46655e0192f28112da2985d28dc728949010730a0ac629dbb7049f9f53a357R2572-R2573)

### CODE REVIEW

**Comment by jfestrada** [comment]
@3d-gussner could you share an example of the `// action:reprint`?

File: Firmware/ultralcd.cpp
Comment (gudnimg) [inline Firmware/ultralcd.cpp]: We can probably save quite a bit of memory by re-using `restore_file_from_sd()` here from one of my PRs: https://github.com/prusa3d/Prusa-Firmware/pull/4437/files

File: Firmware/ultralcd.cpp
Comment (gudnimg) [inline Firmware/ultralcd.cpp]: Please use `enquecommandf_P` and remove `cmd` buffer. This should save some flash memory.
```c
// M23: Select SD file
enquecommandf_P(MSG_M23, altfilename);
// M24: Start/resume SD print
enquecommand_P(MSG_M24);
```

File: Firmware/ultralcd.cpp
Comment (jfestrada) [reply Firmware/ultralcd.cpp]: @gudnimg the only lack about your approach is related with the files which have the _.g_ extension instead of _.gco_
That's something that were already discussed on the original PR #3071, here is the [link for the comments](https://github.com/prusa3d/Prusa-Firmware/pull/3071#discussion_r615726244)
Because of that I added this block I know that it was not the more elegant solution, but it works

	strcpy(altfilename,filename);
	if (!card.FileExists(altfilename))
	{
		strcat_P(filename, PSTR(".gco"));
		if (card.FileExists(filename))
		{
			strcpy(altfilename,filename);
		}else
		{
			strcat_P(altfilename, PSTR(".g"));
		}
	}

File: Firmware/ultralcd.cpp
Comment (jfestrada) [reply Firmware/ultralcd.cpp]: I made that change

Comment (gudnimg) [reply Firmware/ultralcd.cpp]: Nice that’s 56 bytes of flash down ❤️

File: Firmware/ultralcd.cpp
Comment (gudnimg) [reply Firmware/ultralcd.cpp]: I see 🤔 The power panic code should also handle the `.g` extension but it always assume it is `.gco` (Its been that way for years). I think it is time we add the short file name extension into EEPROM. 

We can’t expand `EEPROM_FILENAME` since it will affect adjacent memory. But we can add the extension in a different location. I can fix it in my PR since your PR is unrelated to power panic functionality. 

I think we can reduce the flash increase of your PR from +456 bytes to something closer to +200 bytes. 

We can probably merge your PR first and then later I can create a separate PR to “optimise” it because the flash memory we have left is precious.

Comment (jfestrada) [reply Firmware/ultralcd.cpp]: That make sense, if you need anything from my side just tell me.

File: Firmware/ultralcd.cpp
Comment (gudnimg) [inline Firmware/ultralcd.cpp]: Minor nitpick. But I feel all the calls to `MYSERIAL` in this function should be removed. They increase flash memory usage unnecessarily.

In my opinion they are local debugging code.

```c
MYSERIAL.println(int(depth));
MYSERIAL.println(dir_name);
MYSERIAL.print(altfilename);
```

File: Firmware/ultralcd.cpp
Comment (gudnimg) [inline Firmware/ultralcd.cpp]: Please remove commented code :)

File: Firmware/ultralcd.cpp
Comment (gudnimg) [reply Firmware/ultralcd.cpp]: Resolving this thread for now.

Comment (gudnimg) [reply Firmware/ultralcd.cpp]: I pushed a fix to my PR to support any valid DOS 8.3 file extension. Only tested `".gco"` and `".g"` on my end and it works. I can confirm the current MK3 branch does not work for `".g"`

This did cost ~100 bytes of flash though, but I think it will be worth it since the Re-print functionality can re-use the code.

File: Firmware/ultralcd.cpp
Comment (jfestrada) [reply Firmware/ultralcd.cpp]: It was for debug purpose, already deleted

File: Firmware/ultralcd.cpp
Comment (jfestrada) [reply Firmware/ultralcd.cpp]: I have removed the commented code

File: Firmware/ultralcd.cpp
Comment (gudnimg) [reply Firmware/ultralcd.cpp]: Sorry for picking at these small things but there are two `MYSERIAL` lines left.

Comment (jfestrada) [reply Firmware/ultralcd.cpp]: > Sorry for picking at these small things but there are two `MYSERIAL` lines left.

I'm sorry, I have removed already.

Comment (gudnimg) [reply Firmware/ultralcd.cpp]: Thanks :) Resolving this thread. 

I have yet to test the PR on hardware, will approve later this week if all works as expected.

Comment (jfestrada) [reply Firmware/ultralcd.cpp]: Maybe would be nice to check the translations of "REPRINT" because I used google translator

File: Firmware/messages.cpp
Comment (3d-gussner) [inline Firmware/messages.cpp]: As this is a submenu the allowed char length is `c=18`

File: lang/po/Firmware.pot
Comment (3d-gussner) [inline lang/po/Firmware.pot]: Submenu can be `c=18` long

Comment (3d-gussner) [reply lang/po/Firmware.pot]: `.pot` files and `.po`files have to be updated with `lang/update-pot.py` and `lang/update-po-py`.

**Review by 3d-gussner — CHANGES_REQUESTED** [review]
The language `Firmware.pot` and `Firmware_xx.po` have to be update with the scripts `update-pot.py` and `update.po.py`.
I will change it for you if that is okay.

Submenu  can be 18 character long and NEED to have same `c=` values in the `.po` files as it is in `.pot` file.

**Comment by 3d-gussner** [comment]
The SD card print from the LCD screen is saved to eeprom, but what about the SD card prints started from a host using `M23` and `M24` these aren't covered, or are they?

**Comment by 3d-gussner** [comment]
After a USB/Host print (which is streaming the gcode over serial line) the `Reprint` menu should send a serial message `// action:reprint` to the host. Then the host has to do/run its own sequence depending on this `action`.

We can use the existing  `M118 A1 action:reprint` to send the message `// action:reprint` via serial line.

So the Reprint menu on LCD should:
- restart the SD print
  - When the previous print was a SD print that has been selected via LCD menu before :heavy_check_mark: 
    - We save already the information in eeprom and this PR uses this. That enables the option to reprint also after reset or power cycle
      - I think in this case it would be useful to show which file will be reprinted with an yes/no option to continue with `yes` if times out 
  - When the previous print was a SD print that has been started via `M23` and `M24` gcode commands from USB/host system before :construction: 
     - We don't save this information to eeprom, and not sure if we should do that as well like selecting from LCD screen.
       - We could read from the memory only.  
       - If not implemented the reprint after a reset or power cycle would start the last LCD selected print and not the last via gcode started SD print.
- When the previous print was an USB/Host streamed print send `// action:reprint`. :construction:

File: Firmware/messages.cpp
Comment (jfestrada) [reply Firmware/messages.cpp]: Ok, I suppose that I had to specify the longer length, I will do the change.

**Comment by jfestrada** [comment]
> The SD card print from the LCD screen is saved to eeprom, but what about the SD card prints started from a host using `M23` and `M24` these aren't covered, or are they?

Only if the print is launched from the remote host but with a local SD file, in other case I doubt it should be covered, I mean, when you are printing from SD the main user input is using the LCD, there is where the Reprint function make sense because the user does not need to search the file again on the SD using the LCD menu. When the user is printing from a remote host, their main input is made through that host, so make sense that should be that host which take the control of the Reprint action, usually in octoprint you just need to hit the print button again. Is similar than in mainsail or fluid interfaces and I suppose than in prusaconnect will be the same.
image
Anyway in the case that you are executing the printing from the SD but from a remote host (octoprint, prusaconnect, ...), that should show the reprint menu in the LCD because is similar to execute it from the printer directly, and the _enableReprint_ variable becomes to true when the cmdqueue became empty and the card report the EOF.

```
if(card.eof())
  {
      // file was fully buffered, but commands might still need to be planned!
      // do *not* clear sdprinting until all SD commands are consumed to ensure
      // SD state can be resumed from a saved printing state. sdprinting is only
      // cleared by printingHasFinished after peforming all remaining moves.
      if(!cmdqueue_calc_sd_length())
      {
          // queue is complete, but before we process EOF commands prevent
          // re-entry by disabling SD processing from any st_synchronize call
          card.closefile();
          enableReprint=true;
``` 

Also as the printer is using the SD to print the file the firmware will store the file and the line on eeprom.

**Comment by 3d-gussner** [comment]
@jfestrada The "Reprint" option on LCD is for the User after clearing the bed.

1. Remove print
2. Clean sheet
3. Press Reprint on LCD as they are already there
   a. without having to go back to the computer and press the Reprint button in PrusaLink/PrusaConnect/OctoPrint/other host system.
   b. without having to select the SD again in the dirs etc.

**Comment by 3d-gussner** [comment]
@jfestrada I personally prefer to print from SD and monitor with PrusaLinl/PrusaConnect or Octoprint. When anything happens to the host/network the print continues! And with the RPi port the printer doesn't restart as with USB (you can switch that off BTW) and can continue to monitor the print.

So if I had a farm:

- I would store all files on SD card
- Use PrusaLink/PrusaConnect or Octoprint as central monitoring system.
- Start these SD prints from HOST.

**Comment by jfestrada** [comment]
> After a USB/Host print (which is streaming the gcode over serial line) the `Reprint` menu should send a serial message `// action:reprint` to the host. Then the host has to do/run its own sequence depending on this `action`.
> 
> We can use the existing `M118 A1 action:reprint` to send the message `// action:reprint` via serial line.

OK, I think that could be easy to implement and nice to have.

> So the Reprint menu on LCD should:
> 
>     * restart the SD print
>       
>       * When the previous print was a SD print that has been selected via LCD menu before ✔️
>         
>         * We save already the information in eeprom and this PR uses this. That enables the option to reprint also after reset or power cycle

As it is implemented currently, after a power cycle or eject of the SD the Reprint is disabled, why? because you can't be sure if the last print was OK or if the inserted SD is the same. Would be nice to have the reprint enabled after a power cycle? yes but the firmware will need to check that the last print was finished and that the file exists on the SD before to show the menu for the Reprint. Now the question is, should the reprint be enabled after a powercyle?

>           
>           * I think in this case it would be useful to show which file will be reprinted with an yes/no option to continue with `yes` if times out

Ok, I agree with that

>       * When the previous print was a SD print that has been started via `M23` and `M24` gcode commands from USB/host system before 🚧
>         
>         * We don't save this information to eeprom, and not sure if we should do that as well like selecting from LCD screen.
>           
>           * We could read from the memory only.
>           * If not implemented the reprint after a reset or power cycle would start the last LCD selected print and not the last via gcode started SD print.
currently that will not happen because after reset or powercycle the variable enableReprint is set to false.

Should the reprint be enabled after a powercyle?

>     * When the previous print was an USB/Host streamed print send `// action:reprint`. 🚧
Ok, I agree

**Comment by jfestrada** [comment]
> @jfestrada I personally prefer to print from SD and monitor with PrusaLinl/PrusaConnect or Octoprint. When anything happens to the host/network the print continues! And with the RPi port the printer doesn't restart as with USB (you can switch that off BTW) and can continue to monitor the print.
> 
> So if I had a farm:
> 
>     * I would store all files on SD card
> 
>     * Use PrusaLink/PrusaConnect or Octoprint as central monitoring system.
> 
>     * Start these SD prints from HOST.

That's also my preference, so we have the same point of view

File: Firmware/messages.cpp
Comment (3d-gussner) [reply Firmware/messages.cpp]: The pot file defines all messages and po files are based on it.
That is why we have the two update-po/t.py scripts to keep them correct.
Editing the po/t files manually can break things badly.

**Comment by jfestrada** [comment]
So I will take care of:

* Update language files
* Send the `action:reprint` when the print has been streamed
* Implement the yes/no question with the name of the file to be reprinted in case of a SD print

My question now is, should be enabled the reprint after a power cycle?

**Comment by 3d-gussner** [comment]
> Implement the yes/no question with the name of the file to be reprinted in case of a SD print
> My question now is, should be enabled the reprint after a power cycle?

You made a point that the reprint after power cycle isn't really needed, and so no need to add the extra message.
This was only a though as we have it in eeprom to use it, but as the bool reprint is reset after power cycle i would say no need for additional code and usage of resources.
If users ask for that we can implement it later.

**Comment by arekm** [comment]
> Anyway in the case that you are executing the printing from the SD but from a remote host (octoprint, prusaconnect, ...), that should show the reprint menu

Just to be clear on this - so prints started remotely (from octoprint for example) via M23/M24 are covered and "Reprint" will show up - correct?

**Comment by jfestrada** [comment]
> > Anyway in the case that you are executing the printing from the SD but from a remote host (octoprint, prusaconnect, ...), that should show the reprint menu
> 
> Just to be clear on this - so prints started remotely (from octoprint for example) via M23/M24 are covered and "Reprint" will show up - correct?

Yes, those one are covered with the _Reprint_ function through the LCD menu

**Comment by 3d-gussner** [comment]
@jfestrada Do you have any update commits for this PR?

**Comment by jfestrada** [comment]
> @jfestrada Do you have any update commits for this PR?

I was working on this yesterday (I have made the push just to share the code). Let me explain.

- I have changed the initialization of the _enableReprint_ variable on the cmdqueue because can't be done when the file is closed on the SD, now must be set to true when the buffer is empty, is there any other way to detect the end of a streamed gcode?
- I have added the reference to powerpanic.h on the ultralcd.cpp because I have to know which kind of print I'm doing in order to send the /// action:reprint or just start the reprint from the SD. 
- Now I was dealing with the yes/no question and the name of the file, is there an easy way to reuse the function lcd_show_fullscreen_message_yes_no_and_wait_P with a char[13] as first argument? I do not want to reinvent the wheel.  Is there is no an easy way I will use the function lcd_printf_P.

Also I have to test the action:reprint with octoprint.

**Comment by 3d-gussner** [comment]
@jfestrada Thanks for the update

> Now I was dealing with the yes/no question and the name of the file, is there an easy way to reuse the function lcd_show_fullscreen_message_yes_no_and_wait_P with a char[13] as first argument? I do not want to reinvent the wheel. Is there is no an easy way I will use the function lcd_printf_P.

Let us skip this request as we DON'T reprint from EEPROM after a reboot see above

> You made a point that the reprint after power cycle isn't really needed, and so no need to add the extra message.

**Comment by 3d-gussner** [comment]
> I have added the reference to powerpanic.h on the ultralcd.cpp because I have to know which kind of print I'm doing in order to send the /// action:reprint or just start the reprint from the SD.

We deal with the Host USB printing power panic recovery in a different way, so no need to send `// action:reprint` when power panic recovers.
Or do I miss a case where we should have the reprint option after a power panic?

File: Firmware/messages.cpp
Comment (gudnimg) [reply Firmware/messages.cpp]: Reminder, that `c=7` is still present, should be `c=18`

File: Firmware/cardreader.h
Comment (3d-gussner) [inline Firmware/cardreader.h]: Remove white space

File: Firmware/ultralcd.cpp
Comment (3d-gussner) [inline Firmware/ultralcd.cpp]: Remove white space behind `::PRINT_TYPE_USB)`

**Review by 3d-gussner — CHANGES_REQUESTED** [review]
The PR is working, thanks. Here some changes I have made:
- Rebased to MK3 branch
- Removed white spaces
- Updated Firmware.pot file with ONLY new text (using `update-pot.py `also updates the code lines which isn't useful for a PR. We use it in a pre-release PR to avoid PR conflicts.)
- Looked for existing translations in Prusa-Firmware-Buddy firmware
- Used Deepl and google translate for community languages  

Here the changes for this PR [reprint_pr_changes_diff2.txt](https://github.com/prusa3d/Prusa-Firmware/files/13267653/reprint_pr_changes_diff2.txt)

This is the diff to upstream/MK3 branch [reprint_pr_changes_diff.txt](https://github.com/prusa3d/Prusa-Firmware/files/13267631/reprint_pr_changes_diff.txt)

**Comment by 3d-gussner** [comment]
To test Host/USB `// action:reprint` I installed in OctoPrint the https://plugins.octoprint.org/plugins/actioncommands/ and configured a simple gcode action to send `M117 Reprint`
image

**Comment by 3d-gussner** [comment]
Tested SD reprint:

1. Select SD card print 
2. When finished, choose Reprint from LCD menu starts same SD card file :heavy_check_mark: 
3. After a reset or power up with same SD card inserted the `Reprint` menu starts the last SD card file.

Test USB print:
1. Connect with host and start USB/host print
2. When finished, choose Reprint from LCD menu doesn't send via serial `// action:reprint` :exclamation:
   1. It starts the last SD card print.
3. After a reset or power up with same SD card the `Reprint` menu starts the last SD card print. :exclamation:

**Comment by jfestrada** [comment]
> Tested SD reprint:
> 
> 1. Select SD card print
> 2. When finished, choose Reprint from LCD menu starts same SD card file ✔️
> 3. After a reset or power up with same SD card inserted the `Reprint` menu starts the last SD card file.
> 
> Test USB print:
> 
> 1. Connect with host and start USB/host print
> 2. When finished, choose Reprint from LCD menu doesn't send via serial `// action:reprint` ❗
>    
>    1. It starts the last SD card print.
> 3. After a reset or power up with same SD card the `Reprint` menu starts the last SD card print. ❗

Let me check, this is something related with the initialization of the _enableReprint_ variable, I mean, before to include the USB action, the variable was initialized to true when the card reported the EOF. Now must be set to true when the cmdqueue buffer is empty, but maybe I shall add the check that the printer was printing previously, because after a reboot the _enableReprint_ must be false.

**Review by 3d-gussner — APPROVED** [review]

File: lang/po/Firmware_ro.po
Comment (leptun) [inline lang/po/Firmware_ro.po]: `Repetă print` to be consistent with `Print from SD`

**Review by leptun — CHANGES_REQUESTED** [review]

File: lang/po/Firmware_sk.po
Comment (ingbrzy) [inline lang/po/Firmware_sk.po]: Should be "Vytlačiť znova"

File: lang/po/Firmware_ro.po
Comment (3d-gussner) [reply lang/po/Firmware_ro.po]: @jfestrada Please change it

File: lang/po/Firmware_sk.po
Comment (3d-gussner) [reply lang/po/Firmware_sk.po]: @jfestrada please change it as well

**Comment by jfestrada** [comment]
> > Tested SD reprint:
> > 
> > 1. Select SD card print
> > 2. When finished, choose Reprint from LCD menu starts same SD card file ✔️
> > 3. After a reset or power up with same SD card inserted the `Reprint` menu starts the last SD card file.
> > 
> > Test USB print:
> > 
> > 1. Connect with host and start USB/host print
> > 2. When finished, choose Reprint from LCD menu doesn't send via serial `// action:reprint` ❗
> >    
> >    1. It starts the last SD card print.
> > 3. After a reset or power up with same SD card the `Reprint` menu starts the last SD card print. ❗
> 
> Let me check, this is something related with the initialization of the _enableReprint_ variable, I mean, before to include the USB action, the variable was initialized to true when the card reported the EOF. Now must be set to true when the cmdqueue buffer is empty, but maybe I shall add the check that the printer was printing previously, because after a reboot the _enableReprint_ must be false.

Tested SD reprint:

1. Select SD card print ✔️
2. When finished, choose Reprint from LCD menu starts same SD card file ✔️
3. After a reset or power up with same SD card inserted the `Reprint` menu starts the last SD card file. ✔️

Test USB print:

1. Connect with host and start USB/host print
2. When finished, choose Reprint from LCD menu doesn't send via serial `// action:reprint` ✔️

Tomorrow I will remove the extern bool enableReprint and just use a enableReprint, also I will debug a little bit more, now the enableReprint is set to true on the function _lcdui_print_percent_done_ also I have added an enableReprintUSB in order to control when must be sent an _action:reprint_ to the host.

**Comment by AttilaSVK** [comment]
Hi! The Hungarian translation is correct.

jfestrada ***@***.***> ezt írta (időpont: 2023. nov. 7., K,
16:48):

**Review by leptun — APPROVED** [review]

**Comment by jfestrada** [comment]
> > > Tested SD reprint:
> > > 
> > > 1. Select SD card print
> > > 2. When finished, choose Reprint from LCD menu starts same SD card file ✔️
> > > 3. After a reset or power up with same SD card inserted the `Reprint` menu starts the last SD card file.
> > > 
> > > Test USB print:
> > > 
> > > 1. Connect with host and start USB/host print
> > > 2. When finished, choose Reprint from LCD menu doesn't send via serial `// action:reprint` ❗
> > >    
> > >    1. It starts the last SD card print.
> > > 3. After a reset or power up with same SD card the `Reprint` menu starts the last SD card print. ❗
> > 
> > 
> > Let me check, this is something related with the initialization of the _enableReprint_ variable, I mean, before to include the USB action, the variable was initialized to true when the card reported the EOF. Now must be set to true when the cmdqueue buffer is empty, but maybe I shall add the check that the printer was printing previously, because after a reboot the _enableReprint_ must be false.
> 
> Tested SD reprint:
> 
>     1. Select SD card print ✔️
> 
>     2. When finished, choose Reprint from LCD menu starts same SD card file ✔️
> 
>     3. After a reset or power up with same SD card inserted the `Reprint` menu starts the last SD card file. ✔️
> 
> 
> Test USB print:
> 
>     1. Connect with host and start USB/host print
> 
>     2. When finished, choose Reprint from LCD menu doesn't send via serial `// action:reprint` ✔️
> 
> 
> Tomorrow I will remove the extern bool enableReprint and just use a enableReprint, also I will debug a little bit more, now the enableReprint is set to true on the function _lcdui_print_percent_done_ also I have added an enableReprintUSB in order to control when must be sent an _action:reprint_ to the host.

I have fixed the situation where it was printing from USB, and after that a new print from SD is started, then the reprint was using the one for USB instead of the SD, now that situation is fixed, just setting to false the _enableReprintUSB_ when is printing from SD.

**Comment by 3d-gussner** [comment]
@jfestrada Still have the issue that the Reprint is shown after a reset or power up. A Reprint after an USB print starts the last saved SD print.

**Comment by jfestrada** [comment]
> @jfestrada Still have the issue that the Reprint is shown after a reset or power up. A Reprint after an USB print starts the last saved SD print.

That's doesn't make sense at all, I have tested several times with the new initialization of  the _enableReprint_ variable, are you sure that you are using the last commit from this PR? Now the "Reprint" menu is not enabled after a powercycle or a reboot.

I have made the following test:

1. Send a USB print --> Reprint appears --> restart the printer --> Reprint does not appears
2. Send a SD print --> Reprint appears --> restart the printer --> Reprint does not appears

I also tested: 

1. Send a USB print --> Send a SD print --> Select Reprint --> SD print start to print again
2. Send a SD print --> Send a USB print --> Select Reprint --> Octoprint receive the action:reprint and send the M117 Reprint and appear on the bottom part of the LCD
3. Send a USB print --> Select Reprint --> Octoprint receive the action:reprint and send the M117 Reprint and appear on the bottom part of the LCD
4. Send a SD print --> Select Reprint --> SD print start to print again

**Review by vintagepc — APPROVED** [review]

**Comment by 3d-gussner** [comment]
@Prime1910 @Painkiller56 Please have a look at the translations.

**Comment by 3d-gussner** [comment]
@arekm Please rebase this PR as we have some other changes create a conflict.

**Comment by 3d-gussner** [comment]
@jfestrada Tested the PR and had to fix a minor conflict with current MK3 branch. It is working well with SD prints and Host/USB prints as well.
Please rebase this PR for the last time :pray:

**Comment by jfestrada** [comment]
> @jfestrada Tested the PR and had to fix a minor conflict with current MK3 branch. It is working well with SD prints and Host/USB prints as well. Please rebase this PR for the last time 🙏

I've made the rebase, resolve the conflict accepting the changes that are coming from my branch and should be ready to merge.

**Comment by 3d-gussner** [comment]
@jfestrada Looks like your rebase didn't work correctly. Can you also please update your base branch to the current Prusa MK3 branch as there are so many changed files shown in the PR.

**Comment by 3d-gussner** [comment]
@jfestrada I modified your PR in this diff a little bit after we will merge  #4429 please have a look
[Reprint_after_PR4429.diff.txt](https://github.com/prusa3d/Prusa-Firmware/files/13429851/Reprint_after_PR4429.diff.txt)
The Host Reprint is only shown when `M79` timer is active which means that the host is pining the printer and printer "knows" there is a host.

**Review by ingbrzy — APPROVED** [review]
SK looks OK

**Comment by 3d-gussner** [comment]
Closing in favor of https://github.com/prusa3d/Prusa-Firmware/pull/4510

@jfestrada Thanks for this PR :heart_eyes:

**Pull request closed without merging**
